open-machine-learning-jupyter-book/deep-learning/LSTM.md

```diff
@@ -13,4 +13,635 @@ kernelspec:
+# Reshape each split into [batch_size, training_seq_len]
+batches = [np.resize(x, [batch_size, training_seq_len]) for x in batches]
+
+# Initialize all variables
+init = tf.global_variables_initializer()
+sess.run(init)
+
+# Train model
+train_loss = []
+iteration_count = 1
+for epoch in range(epochs):
+    # Shuffle word indices
+    random.shuffle(batches)
+    # Create targets from shuffled batches
+    targets = [np.roll(x, -1, axis=1) for x in batches]
+    # Run a through one epoch
+    print('Starting Epoch #{} of {}.'.format(epoch+1, epochs))
+    # Reset initial LSTM state every epoch
+    state = sess.run(lstm_model.initial_state)
+    for ix, batch in enumerate(batches):
+        training_dict = {lstm_model.x_data: batch, lstm_model.y_output: targets[ix]}
+        # We need to update initial state for each RNN cell:
+        for i, (c, h) in enumerate(lstm_model.initial_state):
+                    training_dict[c] = state[i].c
+                    training_dict[h] = state[i].h
+        
+        temp_loss, state, _ = sess.run([lstm_model.cost, lstm_model.final_state, lstm_model.train_op],
+                                       feed_dict=training_dict)
+        train_loss.append(temp_loss)
+        
+        # Print status every 10 gens
+        if iteration_count % 10 == 0:
+            summary_nums = (iteration_count, epoch+1, ix+1, num_batches+1, temp_loss)
+            print('Iteration: {}, Epoch: {}, Batch: {} out of {}, Loss: {:.2f}'.format(*summary_nums))
+        
+        # Save the model and the vocab
+        if iteration_count % save_every == 0:
+            # Save model
+            model_file_name = os.path.join(full_model_dir, 'model')
+            saver.save(sess, model_file_name, global_step=iteration_count)
+            print('Model Saved To: {}'.format(model_file_name))
+            # Save vocabulary
+            dictionary_file = os.path.join(full_model_dir, 'vocab.pkl')
+            with open(dictionary_file, 'wb') as dict_file_conn:
+                pickle.dump([vocab2ix, ix2vocab], dict_file_conn)
+        
+        if iteration_count % eval_every == 0:
+            for sample in prime_texts:
+                print(test_lstm_model.sample(sess, ix2vocab, vocab2ix, num=10, prime_text=sample))
+                
+        iteration_count += 1
+
+
+# Plot loss over time
+plt.plot(train_loss, 'k-')
+plt.title('Sequence to Sequence Loss')
+plt.xlabel('Generation')
+plt.ylabel('Loss')
+plt.show()
+```
+
+## Your turn! 🚀
+
+Practice the Long-Short Term Memory Networks by following this TBD.
+
+## Self study
+
+TBD
+
+## Acknowledgments
+
+Thanks to [Nick](https://github.com/nfmcclure) for creating the open-source course [tensorflow_cookbook](https://github.com/nfmcclure/tensorflow_cookbook) and [Sebastian Raschka](https://github.com/rasbt) for creating the open-sourse [stat453-deep-learning-ss20](https://github.com/rasbt/stat453-deep-learning-ss20). It inspires the majority of the content in this chapter.

```

Comment:
It inspires -> They inspire

Reply:
open-sourse -> open-source

Reply:
is the tensorflow_cookbook a course? is the stat453-deep-learning-ss20 a course. Please correct the statement.

Reply:
Done. "Course" has been changed to "library".